spec/src/main/asciidoc/core/invokers.asciidoc

```diff
@@ -0,0 +1,154 @@
+However, during application deployment, frameworks may observe application methods through CDI extensions and build an `Invoker` for each relevant method.
+The invokers can then be used at application runtime to invoke the methods indirectly.
+
+Method invokers are not supposed to be used by application code, as applications may invoke their own methods directly.
+
+[[building_invoker]]
+=== Building an `Invoker`
+
+The CDI container allows building an `Invoker` for a method of an <<enablement,enabled>> <<managed_beans,managed bean>>.
+The method for which the invoker is built is called the _target method_ of the invoker, and the managed bean is called the _target bean_ of the invoker.
+
+Invalid target methods are:
+
+* private methods,
+* constructors,
+* methods declared on the `java.lang.Object` class, except of the `toString()` method,
+* methods that are not declared on the bean class of the target bean or inherited from its supertypes.
+
+Attempting to build an invoker for an invalid target method leads to a deployment problem.
+
+When the target bean is not a managed bean, attempting to build an invoker leads to a deployment problem.
+When the target bean is an interceptor, attempting to build an invoker leads to a deployment problem.
+
+Multiple managed beans may inherit a method from a common supertype.
+In that case, an invoker must be built for each target bean individually.
+An invoker built for one target bean may not be used to invoke the target method on an instance of another target bean.
+
+The only way to build an invoker is using the `InvokerBuilder`.
+An `InvokerBuilder` can only be obtained in CDI portable extensions and build compatible extensions.
+See <<invoker_builder>> for more information.
+
+[[using_invoker]]
+=== Using an `Invoker`
+
+The `Invoker` interface contains a single method:
+
+[source,java]
+----
+public interface Invoker<T, R> {
+    R invoke(T instance, Object[] arguments) throws Exception;
+}
+----
+
+Calling `invoke()` invokes the target method on given `instance` of the target bean, passing given `arguments`.
+
+Invoker implementations must be thread-safe.
+A single invoker instance may be used to perform multiple, possibly concurrent, invocations of the target method, possibly on different instances of the target bean, possibly with different arguments.
+
+Whenever a direct invocation of a method on an object is a business method invocation, an indirect invocation of that method on that object through an invoker is also a business method invocation.
+
+==== Behavior of `invoke()`
+
+If the target method is `static`, the `instance` is ignored; by convention, it should be `null`.
+If the target method is not `static` and `instance` is `null`, a `RuntimeException` is thrown.
+If the target method is not `static` and the class of the `instance` is not a subclass of the bean class of the target bean, a `RuntimeException` is thrown.
```

Comment:
> the class of the `instance` is not a subclass of the bean class of the target bean

We probably need to relax this somehow, because contextual references (and therefore injectable references) don't necessarily need to satisfy this constraint. Not yet sure...

Reply:
I tried to clarify the requirements on the target instance in the last commit, but it probably needs more thinking.

I was originally convinced it should be enough to demand that the target instance is an instance of the bean class, but there are rules (https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0.html#typecasting_between_bean_types) that make this more complex (I vaguely recall something to do with EJBs too).

Reply:
I think I figured this out, added one more commit. The gist is:

We already define that the target method must be declared by the bean class of the target bean, or inherited from a supertype. That remains, and failure is still a deployment problem (basically a bean/method mismatch). We however also define that the target method should be declared by a type that is present in the set of bean types of the target bean. If not, it is a non-portable behavior, since some implementations can support it. (As an aside, I also made it non-portable to register an invoker for a `final` method. Again, some implementations can support it.)

This lets us mandate minimum requirements on the target instance in case it is a contextual reference for the target bean (likely the most common case), in line with the rest of the specification. If the target method is declared on an interface, the contextual reference can always be cast to it, otherwise the contextual reference must be for the type that declares the target method. For a non-contextual instance, the class simply must declare the method or inherit from a supertype. Implementations may support more.

Took me a while to figure out, but I feel pretty happy about it.

Reply:
> As an aside, I also made it non-portable to register an invoker for a `final` method. Again, some implementations can support it.

Added one more commit where I expanded this to all target methods declared on unproxyable bean types.

My reasoning was that if a bean type is unproxyable, it does not need to be possible to obtain a contextual reference for a bean for that bean type, and invoking a method through a contextual reference is likely to be the most common case, so it's better to be open about it.

I however realize that this might be too broad. There are 3 cases when CDI cares about proxyability:

- normal scoped beans
- beans with bound interceptors
- beans with bound decorators

In the 1st case, CDI mandates that all bean types for which a contextual reference is ever obtained are proxyable. In the 2nd and 3rd case, CDI mandates just the _bean class_ to be a proxyable type.

I hope we can tighten the requirement here somehow, but I'm not sure how much.